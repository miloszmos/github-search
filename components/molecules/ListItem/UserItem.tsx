import Link from 'next/link';
import React from 'react';
import { User } from '../../../graphql/generated/graphql';
import Avatar from '../../atoms/Avatar/Avatar';
import { LabelStyles } from '../../atoms/Label/Label.styles';
import {
  ItemDescriptionStyles,
  ItemTextStyles,
  ItemWrapperStyles,
} from './ListItem.styles';

interface UserItemProps {
  user: User;
}

const UserItem = ({ user }: UserItemProps) => {
  const { avatarUrl, name, login, location, bio } = user;
  return (
    <ItemWrapperStyles>
      <div>
        <Avatar src={avatarUrl} alt="some nice faces" />
      </div>
      <div className="right">
        {name && (
          <Link href={`/user/${login}`}>
            <a>
              <ItemTextStyles className="mb">{name}</ItemTextStyles>
            </a>
          </Link>
        )}
        <ItemTextStyles className="gray">{login}</ItemTextStyles>
        <ItemDescriptionStyles>{bio}</ItemDescriptionStyles>
        <LabelStyles>{location}</LabelStyles>
      </div>
    </ItemWrapperStyles>
  );
};

export default UserItem;
