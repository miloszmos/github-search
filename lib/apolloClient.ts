import { ApolloClient, InMemoryCache, createHttpLink } from '@apollo/client';
import { setContext } from '@apollo/client/link/context';

const httpLink = createHttpLink({
  uri: process.env.NEXT_PUBLIC_API_ENDPOINT,
});

const authLink = setContext((_, { headers }) => ({
  headers: {
    ...headers,
    authorization: `Bearer ${process.env.NEXT_PUBLIC_ACCESS_TOKEN}`,
  },
}));

export const client = new ApolloClient({
  link: authLink.concat(httpLink),
  cache: new InMemoryCache(),
});
