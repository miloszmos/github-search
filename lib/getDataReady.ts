import { User, Repository, SearchQuery } from '../graphql/generated/graphql';

type ResponseDataType = (Repository | User | null)[];

export const getDataReady = (data: SearchQuery): ResponseDataType => {
  const repoNodes = data.repositories.nodes ?? [];
  const userNodes = data.users.nodes ?? [];
  const mergedData = [...repoNodes, ...userNodes].filter(Boolean) as (
    | Repository
    | User
  )[];

  return mergedData
    .filter(
      // ? Check for User | Repository because I had some other types in response
      (el) => el.__typename && el.__typename.match(/^(User|Repository)$/)
    )
    .sort((a, b) => a.id.localeCompare(b.id));
};
