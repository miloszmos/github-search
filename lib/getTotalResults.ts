import { SearchQuery } from '../graphql/generated/graphql';

export const getTotalResults = (data: SearchQuery): number => {
  return data.repositories.repositoryCount + data.users.userCount;
};
