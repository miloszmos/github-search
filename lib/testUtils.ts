import casual from 'casual';

casual.seed(777);

const fakeUser = () => ({
  __typename: 'User',
  name: casual.name,
  avatarUrl: casual.url,
  bio: casual.description,
  id: casual.uuid,
  location: casual.address,
  login: casual.username,
});

const fakeRepository = () => ({
  __typename: 'Repository',
  id: casual.uuid,
  nameWithOwner: casual.full_name,
  createdAt: '2022-12-11T20:00:00.797Z',
  description: casual.description,
  stargazerCount: casual.random,
  url: casual.url,
  updatedAt: '2022-12-11T20:00:00.797Z',
  licenseInfo: {
    name: casual.name,
  },
  languages: [
    {
      nodes: {
        name: casual.name,
        color: casual.color_name,
      },
    },
  ],
});

export { fakeUser, fakeRepository };
