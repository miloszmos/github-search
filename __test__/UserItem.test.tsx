import React from 'react';
import { render, screen } from '@testing-library/react';
import { MockedProvider } from '@apollo/react-testing';
import { fakeUser } from '../lib/testUtils';
import UserItem from '../components/molecules/ListItem/UserItem';

const item = fakeUser();

describe('<UserItem />', () => {
  it('renders and matches snapshot', () => {
    const { container } = render(
      <MockedProvider>
        <UserItem user={item} />
      </MockedProvider>
    );
    expect(container).toMatchSnapshot();
  });

  it('renders out base information', async () => {
    const { container } = render(
      <MockedProvider>
        <UserItem user={item} />
      </MockedProvider>
    );
    const nameWithOwner = screen.getByText(item.name);
    expect(nameWithOwner).toBeInTheDocument();
    const img = screen.getByAltText('some nice faces');
    expect(img).toBeInTheDocument();
  });
});
