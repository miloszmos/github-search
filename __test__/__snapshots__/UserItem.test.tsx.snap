// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<UserItem /> renders and matches snapshot 1`] = `
<div>
  <div
    class="sc-b08aa592-2 kghpTT"
  >
    <div>
      <span
        style="box-sizing: border-box; display: inline-block; overflow: hidden; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px; position: relative; max-width: 100%;"
      >
        <span
          style="box-sizing: border-box; display: block; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px; max-width: 100%;"
        >
          <img
            alt=""
            aria-hidden="true"
            src="data:image/svg+xml,%3csvg%20xmlns=%27http://www.w3.org/2000/svg%27%20version=%271.1%27%20width=%2720%27%20height=%2720%27/%3e"
            style="display: block; max-width: 100%; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px;"
          />
        </span>
        <img
          alt="some nice faces"
          aria-label="user avatar"
          class="sc-6cb0753b-0 btFnbM"
          data-nimg="intrinsic"
          decoding="async"
          src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7"
          style="position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px; box-sizing: border-box; padding: 0px; margin: auto; display: block; width: 0px; height: 0px; min-width: 100%; max-width: 100%; min-height: 100%; max-height: 100%;"
        />
        <noscript />
      </span>
    </div>
    <div
      class="right"
    >
      <a
        href="/user/Kevin.Schimmel"
      >
        <p
          class="sc-b08aa592-0 wvFcc mb"
        >
          Miss Coleman Berge
        </p>
      </a>
      <p
        class="sc-b08aa592-0 wvFcc gray"
      >
        Kevin.Schimmel
      </p>
      <p
        class="sc-b08aa592-1 eptwLV"
      >
        Laborum est possimus assumenda aut enim iure facilis ut. Iure vitae occaecati officia. Nisi ut maiores recusandae fuga reprehenderit. Non omnis consequatur enim quia autem et modi.
      </p>
      <span
        class="sc-d9d70f80-0 crKzRe"
      >
        2073 Greenholt Springs Suite 265
South Odessa, AZ 88118
      </span>
    </div>
  </div>
</div>
`;
